src/main/java/com/uid2/admin/vertx/service/SharingService.java

```diff
@@ -79,21 +79,33 @@ private void handleSetKeyset(RoutingContext rc) {
                 rc.fail(500);
             }
 
+            final Map<Integer, Keyset> keysetsById = this.keysetProvider.getSnapshot().getAllKeysets();
+
             final JsonObject body = rc.body().asJsonObject();
 
             final JsonArray whitelist = body.getJsonArray("allowed_sites");
             Integer keysetId = body.getInteger("keyset_id");
             Integer siteId = body.getInteger("site_id");
             String name = body.getString("name", "");
 
-            if (keysetId == null && siteId == null
-                    || keysetId != null && siteId != null) {
+            if ((keysetId == null && siteId == null)
+                    || (keysetId != null && siteId != null)) {
                 ResponseUtil.error(rc, 400, "You must specify exactly one of: keyset_id, site_id");
                 return;
             }
+            Integer finalSiteId = siteId;
+            if (siteId != null &&
+                    (siteId == Const.Data.AdvertisingTokenSiteId
+                            || siteId == Const.Data.RefreshKeySiteId
+                            || siteId == Const.Data.MasterKeySiteId
+                            || siteProvider.getSite(siteId) == null
+                            || keysetsById.values().stream().anyMatch(k -> k.getSiteId() == finalSiteId))) {
```

Comment:
I don't think we need this line, if a site is not in the site Provider it won't be in the the keysets. 
I.E. that means we would have previously added a site that wasn't in the site provider, which is impossible because we check the site provider each time .

Reply:
if I remove the last check then if the site is in the site provider it is considered valid, even if there is already a keyset for it. If I remove the second to last check then we can create only one keyset but for any site number even if it's not in the site provider.

Reply:
Just add a comment that this is now to enforce the single keyset

Reply:
commented